### PR TITLE
Fix mining in RPC tests

### PR DESCRIPTION
**Pull request by gumb0**
This adds to the `Client` ability to send callbacks about mined block and imported block, so that the unit tests can be notified and keep exact track of how many blocks were generated, instead of waiting in the loop.
(This central change is in TestHelper.cpp)

Also some minor changes, e.g. fixing RPC calls for "latest" block.

### CODE REVIEW

**Comment by winsvega** [comment]
could you run this comman locally: 
```
./testeth -t BlockchainTests/bcStateTests -- --filltests
```

and see if its generating the tests correctly it should return 100% and no errors.

File: libethereum/Client.cpp [hunk 1]
```diff
@@ -615,11 +621,14 @@ void Client::rejigSealing()
 
             if (wouldSeal())
             {
-                sealEngine()->onSealGenerated([=](bytes const& header){
-                    if (!this->submitSealed(header))
+                sealEngine()->onSealGenerated([=](bytes const& header) {
+                    LOG(m_logger) << "Block seal generated";
```
Comment (chfast) [inline libethereum/Client.cpp]: `"Block sealed"`?

**Comment by gumb0** [comment]
@winsvega this command finishes for me with
```
.
C:/dev/cpp-ethereum/test/tools/libtesteth/TestHelper.cpp(444): error: in "BlockchainTests/bcStateTests": executeCmd() has not been implemented for Windows.
C:/dev/cpp-ethereum/test/tools/libtesteth/TestHelper.cpp(444): error: in "BlockchainTests/bcStateTests": executeCmd() has not been implemented for Windows.
C:/dev/cpp-ethereum/test/tools/libtesteth/TestHelper.cpp(444): error: in "BlockchainTests/bcStateTests": executeCmd() has not been implemented for Windows.
C:/dev/cpp-ethereum/test/tools/libtesteth/TestHelper.cpp(444): error: in "BlockchainTests/bcStateTests": executeCmd() has not been implemented for Windows.
C:/dev/cpp-ethereum/test/tools/libtesteth/TestHelper.cpp(444): error: in "BlockchainTests/bcStateTests": executeCmd() has not been implemented for Windows.
22%...
C:/dev/cpp-ethereum/test/tools/libtesteth/TestHelper.cpp(144): error: in "BlockchainTests/bcStateTests": blockhashNonConstArg network not found: ALL
C:/dev/cpp-ethereum/test/tools/libtesteth/TestHelper.cpp(474): error: in "BlockchainTests/bcStateTests": LLL compilation only supported on posix systems.
C:/dev/cpp-ethereum/test/tools/libtesteth/TestHelper.cpp(358): fatal error: in "BlockchainTests/bcStateTests": Bytecode is missing! '{ [[ 1 ]] (BLOCKHASH (BALANCE 0x1000000000000000000000000000000000000000)) }' blockhashNonConstArg_Frontier

*** 8 failures are detected (5 failures are expected) in the test module "Master Test Suite"
```
which seems to be windows-specific.

Actually I think blockhcain tests use another `mine()` function which I didn't touch.

**Comment by gumb0** [comment]
@chfast I don't think it's related, the function I fix here is used only in RPC tests, not in private networks.

Do we have an issue for that? I could debug it.

**Comment by chfast** [comment]
> Do we have an issue for that? I could debug it.

To be reported soon.

File: libethereum/Client.cpp (hunk 1 above)
Comment (gumb0) [reply libethereum/Client.cpp]: Well right after this message there goes (on trace verbosity) "Sealing block!" https://github.com/ethereum/cpp-ethereum/blob/dacd74b4d0bba7ef131b19726e4346662f3dd4de/libethereum/Block.cpp#L840

So now it's like seal generated, then this seal is appended to the block.

Comment (chfast) [reply libethereum/Client.cpp]: Yes, I've seen also "Generating seal..." which sounds better than "Sealing..."

Comment (gumb0) [reply libethereum/Client.cpp]: Yeah, I think I'll just delete "Sealing block!" which is confusing

File: test/tools/libtesteth/TestHelper.cpp [hunk 2]
```diff
@@ -43,17 +43,24 @@ namespace dev
...
+    int importedBlocks = 0;
+    std::promise<void> allBlocksImported;
+    auto importHandler =
+        _c.setOnBlockImport([_numBlocks, &importedBlocks, &allBlocksImported](BlockHeader const&) {
+            if (++importedBlocks == _numBlocks)
+                allBlocksImported.set_value();
+        });
+
+    _c.startSealing();
+    allBlocksImported.get_future().get();
```
Comment (chfast) [inline test/tools/libtesteth/TestHelper.cpp]: Replace `.get()` with `.wait()`?

File: libethereum/Client.h [hunk 3]
```diff
@@ -264,6 +281,7 @@ class Client: public ClientBase, protected Worker
 
     /// Called after processing blocks by onChainChanged(_ir)
     void resyncStateFromChain();
+    void restartMining();
```
Comment (chfast) [inline libethereum/Client.h]: A documentation comment needed?

**Comment by winsvega** [comment]
@gumb0  sorry. this command require solidity

**Comment by gumb0** [comment]
@winsvega can you give me the command to generate just one blockchain test, I can try to compare it on my machine

**Comment by winsvega** [comment]
blockchain tests has issues. let me commit to this branch with a fix

**Comment by gumb0** [comment]
Well this is not related to blockchain tests, so maybe better another PR?

**Comment by winsvega** [comment]
ok. you could run the test command locally with my PR
```./testeth -t BlockchainTests/bcStateTests --  --all --singletest blockhashNonConstArg --verbosity 10```

but remove the PR before merging. I'll open the new PR

**Comment by winsvega** [comment]
here the separate PR:
https://github.com/ethereum/cpp-ethereum/pull/5092/files

you could remove my last commit from here

**Comment by gumb0** [comment]
Actually we should have measured blockhain tests filling performance when we changed ethash library

**Comment by gumb0** [comment]
@winsvega Well I didn't get lllc to work, but current develop doesn't seem faster for me, it also often displays 5 or more "." when mining in this command.

Could it be that you compare not with current develop but the version from while ago? We replaced libethash ~month ago https://github.com/ethereum/cpp-ethereum/pull/5051 and it could change performance in your environment

**Pull request merged**